<a href="https://colab.research.google.com/github/ajakrish/Ajay_info5502_spring2022/blob/main/lab_assignment_03_(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## The third Lab-assignment (02/10/2022, 50 points in total)

The purpose of this exercise is to understand users' information needs, then collect data from different sources for analysis.

Question 1 (10 points). Fomulate your domain problem: Describe an interesting research question (or practical question) you have in mind, what kind of data should be collected to answer the question(s)? How many data needed for the analysis? The detail steps for collecting and save the data. 

In [ ]:
# Your answer here (no code for this question, write down your answer as detail as possible for the above questions):

'''
Please write you answer here:
Domain Problem:
# Domain Problem is to analyse the rating given by customers for phones in flipkart website.

Research question:
# Do the rating data analysis

Data Required to answer the question:
The following are the attributes of the required data to answer the research question
# We need mobile phone data attribute and ratings given by a customer for a particular mobile phone.

We can get 1000 results of mobile phones from the following link: https://www.flipkart.com/search?q=phone&sid=tyy%2C4io&as=on&as-show=on&otracker=AS_QueryStore_OrganicAutoSuggest_1_4_na_na_na&otracker1=AS_QueryStore_OrganicAutoSuggest_1_4_na_na_na&as-pos=1&as-type=RECENT&suggestionId=phone%7CMobiles&requestId=1d2042af-103c-407c-a281-8a902ea36e9c&as-searchtext=phon

Steps for collecting and saving the data:
. To scrape the data we use web scraping technique.
. To parse the Flipkart website we use BeautifulSoup library .
. We parse through each page from range 1 to 100 and we go to the particular mobile phone link.
. Then we got 1000 results of mobile phones from 24 pages.
. We get the rating from a particular mobile phone.
. The rating details are obtained from the transcript hyperlink.
. We parse the Segment Body to Paragraph elements in the html code.
. Then we store the data in a json format
. Finally, data is prepared(json format) for our research.






'''

Question 2 (10 points). Collect your data to answer the research problem: Write python code to collect 1000 data samples you discussed above.

In [2]:
# You code here (Please add comments in the code):
from bs4 import BeautifulSoup
import urllib
from urllib.request import Request, urlopen
from urllib.error import HTTPError
import json
import re

counts = 0

ratings = []
main_url = "https://www.flipkart.com/search?q=phone&sid=tyy%2C4io&as=on&as-show=on&otracker=AS_QueryStore_OrganicAutoSuggest_1_4_na_na_na&otracker1=AS_QueryStore_OrganicAutoSuggest_1_4_na_na_na&as-pos=1&as-type=RECENT&suggestionId=phone%7CMobiles&requestId=1d2042af-103c-407c-a281-8a902ea36e9c&as-searchtext=phon"
for page_num in range(1, 100):
    # we create the data soup of the main url page html elements.
    link1 = Request(main_url.format(page_num), headers={'User-Agent': 'Mozilla/5.0'})
    url1 = urllib.request.urlopen(link1)

    data1 = url1.read()
    data1_soup = BeautifulSoup(data1)

    print("*** Page {} ***".format(page_num))
    # we iterate through pages of the  search results till we get 1000 records...
    for link in data1_soup.find_all('div', attrs = {'class': "_3LWZlK"}):
        ratings.append(link.text)
        counts = counts + 1
        if counts == 1000:
            break
            
    print(counts)
    if counts == 1000:
        break
    
print(ratings)







*** Page 1 ***
39
*** Page 2 ***
78
*** Page 3 ***
117
*** Page 4 ***
156
*** Page 5 ***
195
*** Page 6 ***
234
*** Page 7 ***
273
*** Page 8 ***
312
*** Page 9 ***
351
*** Page 10 ***
390
*** Page 11 ***
429
*** Page 12 ***
468
*** Page 13 ***
507
*** Page 14 ***
546
*** Page 15 ***
585
*** Page 16 ***
624
*** Page 17 ***
663
*** Page 18 ***
702
*** Page 19 ***
741
*** Page 20 ***
780
*** Page 21 ***
819
*** Page 22 ***
858
*** Page 23 ***
897
*** Page 24 ***
936
*** Page 25 ***
975
*** Page 26 ***
1000
['4.4', '4.5', '4.5', '4.6', '4.6', '4.3', '4.3', '4.3', '4.3', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.3', '4.4', '4.4', '4.5', '4.5', '5', '1', '4.4', '4', '3', '4.4', '4', '5', '4.6', '3', '4', '4.4', '4', '1', '4.4', '4.6', '4.6', '4.5', '4.5', '4.3', '4.3', '4.3', '4.3', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.4', '4.3', '4.4', '4.5', '4.5', '5', '1', '4.4', '4', '3', '4.4', '4', '5', '4.6', '3', '4', 

Question 3 (10 points). Understand the data quality: Search a second hand dataset (any dataset) from kaggle or other websites. Describe the data quality problem of the dataset and explain your strtegy to clean the data.

In [ ]:
# Your answer here (no code for this question, write down your answer as detail as possible for the above questions):
'''

Please write you answer here:
Dataset:
I have taken a pakistan suicide attacks dataset from kaggle and the dataset link is: https://www.kaggle.com/zusmani/pakistansuicideattacks
The dataset has 26 columns and 496 rows of data.
The datset contains missing values so I have some approaches to clean the data:
Using Python Library we need to find which cell has missing values.
After finding the missing values cell:
1. First Approach is used to drop the columns from dataset. It is the best solution if most of the data from that column is missing.
2. Second Approach is Imputation. It fills the missing value with some number. The value won't be same in most cases, but in most cases it gives accurate models than dropping the entire column.
3. Third approach is Extension to impuatation. The imputed values may be below or above the actual values. The rows with missing values may be unique in some way. In this case, model will make better predictions by considering all the missing values.




'''

Question 4 (20 points). Data cleaning: There are two datasets TwADR-L (from Twitter) and AskAPatient (Link: https://zenodo.org/record/55013#.YgU2NN-ZO4T) for medical concept
normalization. However, the two datasets have serious data quality problems. Please read the introduction of the datasets and clean the two datasets by following the steps in the statement.

In the original dataset, the TwADR-L had 48,057 training, 1,256 validation and 1,427 test examples. The test set (all
test samples from 10 folds combined) consists of 765 unique phrases and 273 unique classes (medical concepts). The AskAPatient dataset contained 156,652 training, 7,926 validation, and 8,662 test examples. The entire test set (all test samples
from 10 folds combined) consists of 3,749 unique phrases and 1,035 unique classes (medical concepts). The authors
randomly split each dataset into ten equal folds, ran 10-fold cross validation and reported the accuracy averaged across the
ten folds. 

We found that, in the original data set, many phrase-label pairs appeared multiple times within the same training data file
and also across the training and test data sets in the same fold. In the AskAPatient data set, on average 35.82% of the test data overlapped with training data in the same fold. In the Twitter (TwADR-L) dataset, on average 8.62% of the test set had an overlap with the training data in the same fold. Having a large overlap between the training and the test data can potentially
introduce bias in the model and contribute to high accuracy. It is not unlikely that the high model performance reported in the original paper may be triggered by the the large overlap between the training and test sets.

Therefore to remove the bias, we further cleaned and recreated the training, validation, and test sets such that each
phrase-label pair appears only once in the entire dataset (either in training, validation or test set).

(1) First, we combined all examples in training, validation and test data from the original data set and then removed all
duplicate phrase-label pairs (examples that have the same phrase and label pair and appear more than once in training/validation/test datasets). Table II shows the statistics of the new dataset (after removing duplicates). The Twitter data set had 3,157 unique phrase-label pairs and 2,220 unique labels (medical concepts) while 173 phrases had multiple labels (i.e., they were assigned to more than one label). Many concepts had only one example, and the concept that had the most number
of examples had 36 phrases. On average, each concept had 1.42 examples. The AskAPatient data set had 4,496 unique phrase-label pairs, 1,036 unique labels while 26 phrases had multiple labels. Table III shows examples of phrases that had multiple labels. For example, ‘mad’ can be mapped to ‘anger’ or ‘rage’ and ‘sore’ can be mapped to ‘pain’ or ‘myalgia’.

(2) Second, we remove all concepts that had less than five examples. The statistics of the final data are shown in Table IV.

(3) Third, we divide all examples without multiple labels into random 10 folds such that each unique phrase-label pair
appears once in one of the 10 test sets. We add the pairs with multiple labels into the training data. This final 10-folds
dataset is used in all our experiments.

(The original paper can be download on canvas)

In [9]:
# You code here (Please add comments in the code):
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [7]:
import pandas as pd
import glob
result_df =  pd.DataFrame()
# we iterate through both datasets 
for dataset in ['TwADR-L', 'AskAPatient']:
    result = {}
    df = pd.DataFrame()
    for filepath in glob.glob('/content/drive/MyDrive/datasets/{}/*.txt'.format(dataset)):
        data = pd.read_csv(filepath, sep = "\t", header = None, encoding= 'unicode_escape')
        df = df.append(data)
    # we create combined dataframe by combining Each dataset text files.
    df = df.reset_index(drop=True)
    df['phrase_label'] = df[1] + " " + df[2]
    df.columns = ['id', 'labels', 'phrases', 'phrase-label']
    # type casting to str
    df = df.astype({"id": str})
    # converting all values to lower case
    for columns in df.columns:
        df[columns] = df[columns].str.lower() 

    # dropping duplicates
    df = df.drop_duplicates('phrase-label')

    # storing results of the table in dictionary
    result['Unique_phrases'] = len(df['phrases'].unique())
    result['Unique_labels'] = len(df['labels'].unique())
    result['Unique_phrase_label_pairs'] = df.shape[0]
    df1 = pd.DataFrame(df['phrases'].value_counts())
    result['Phrases with multiple labels'] = df1[df1['phrases'] > 1].shape[0]
    result['Min examples per label'] = df['labels'].value_counts().values.min()
    result['Max examples per label'] = df['labels'].value_counts().values.max()
    result['Avg examples per label'] = round(df['labels'].value_counts().mean(), 2)
    result_df = result_df.append(result, ignore_index=True)


result_df = result_df.T
#result_df = result_df.astype({0: int, 1: int})
result_df.columns = ['TwADR-L', 'AskAPatient']

result_df

TwADR-L  AskAPatient
Unique_phrases                2944.00      4470.00
Unique_labels                 2220.00      1038.00
Unique_phrase_label_pairs     3157.00      4507.00
Phrases with multiple labels   173.00        35.00
Min examples per label           1.00         1.00
Max examples per label          36.00       141.00
Avg examples per label           1.42         4.34

In [8]:
import pandas as pd
import glob
result_df = pd.DataFrame()
# we iterate through both datasets 
for dataset in ['TwADR-L', 'AskAPatient']:
    df = pd.DataFrame()
    result = {}
    for filepath in glob.glob('/content/drive/MyDrive/datasets/{}/*.txt'.format(dataset)):
        #print(filepath)
        data = pd.read_csv(filepath, sep = "\t", header = None, encoding= 'unicode_escape')
        df = df.append(data)
    # we create combined dataframe by combining Each dataset text files.
    df = df.reset_index(drop=True)
    df['phrase_label'] = df[1] + " " + df[2]
    df.columns = ['id', 'labels', 'phrases', 'phrase-label']
    #print(df.shape)
    df = df.astype({"id": str})
    # converting to lower case
    for columns in df.columns:
        df[columns] = df[columns].str.lower() 

    # we drop duplicates that are phrase-label pairs... 
    df = df.drop_duplicates('phrase-label')

    index_list = []
    for i in range(df.shape[0]):
        if df['labels'].value_counts()[df.iloc[i]['labels']] < 5:
            index_list.append(i)

    # we drop the labels that have less than count of 5.
    df.drop(df.index[index_list], inplace=True)
    result = {}
    # storing results of the table in dictionary
    result['Unique_phrases'] = len(df['phrases'].unique())
    result['Unique_labels'] = len(df['labels'].unique())
    result['Unique_phrase_label_pairs'] = df.shape[0]
    df1 = pd.DataFrame(df['phrases'].value_counts())
    result['Phrases with multiple labels'] = df1[df1['phrases'] > 1].shape[0]
    result['Min examples per label'] = df['labels'].value_counts().values.min()
    result['Max examples per label'] = df['labels'].value_counts().values.max()
    result['Avg examples per label'] = round(df['labels'].value_counts().mean(), 2)
    # appending dictinary to dataframe
    result_df = result_df.append(result, ignore_index=True)

result_df = result_df.T
#result_df = result_df.astype({0: str, 1: str})
result_df.columns = ['TwADR-L', 'AskAPatient']

result_df

TwADR-L  AskAPatient
Unique_phrases                 616.00      2665.00
Unique_labels                   76.00       233.00
Unique_phrase_label_pairs      721.00      2686.00
Phrases with multiple labels    87.00        19.00
Min examples per label           5.00         5.00
Max examples per label          36.00       141.00
Avg examples per label           9.49        11.53